DeeperFunc/Router.swift

```diff
@@ -0,0 +1,206 @@
+//
+//  Router.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 26/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+public protocol Route: Equatable {
+    func deconstruct<A>(_ constructor: ((A) -> Self)) -> A?
+}
+
+public class Router<U: Route>: DeepLinkRouter, CustomStringConvertible {
+    public let scheme: String
+    public let rootDeepLinkHandler: AnyDeepLinkHandler<U>?
+    
+    private var route: RoutePattern<U, Path>?
+
+    public init(scheme: String, rootDeepLinkHandler: AnyDeepLinkHandler<Intent>) {
+        self.scheme = scheme
+        self.rootDeepLinkHandler = rootDeepLinkHandler
+    }
+
+    public var description: String {
+        return "\(type(of: self)):\n[\(route?.template ?? "")]"
+    }
+    
+    public func openURL(_ url: URL) -> U? {
+        guard
+            let components = url.routeComponents,
+            url.scheme == scheme,
+            let result = route?.parse(components),
+            result.rest.path.isEmpty
+            else { return nil }
+        
+        return result.match
+    }
+
+    private func add(_ route: RoutePattern<U, Path>) -> Router {
+        self.route = self.route.map({ oldValue in
+            .init(parse: { oldValue.parse($0) ?? route.parse($0) },
+                  print: { oldValue.print($0) ?? route.print($0) },
+                  template: "\(oldValue.template)\n\(route.template)")
+        }) ?? route
+        return self
+    }
+    
+    @discardableResult
+    public func add(_ intent: U, route: String) -> Router {
+        return add(intent, route: lit(route))
+    }
+    
+    @discardableResult
+    public func add<S: ClosedPatternState>(_ intent: U, route: RoutePattern<Void, S>) -> Router {
+        return add(route.map({ intent }, { $0 == intent ? () : nil }))
+    }
+    
+    @discardableResult
+    public func add(_ intent: U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<Void, Path> = route.map({ _ in () }, { _ in () })
+        return add(intent, route: typedRoute)
+    }
+
+    @discardableResult
+    public func add<A, S: ClosedPatternState>(_ intent: @escaping ((A)) -> U, route: RoutePattern<A, S>) -> Router {
+        return add(route.map({ intent($0) }, { $0.deconstruct(intent) }))
```

Comment:
All routes are mapped to the same generic type, `<U, Path>`, so that they can be composed into one route.
Input parameter of second closure will be value of `U`, but `router` needs `A`, which is associated value of enum case (if enum is used of course). But we only have constructor of `U`, so we don't know with what case to match the value. This is done instead with `deconstruct` method that does pattern matching and returns associated values.
This is only needed to be able to print urls for specific values of `U`. We could print just using route itself (see this commit), but we can't get it back from router, so will need to store all routes in some constants